azkaban-common/src/main/java/azkaban/imagemgmt/permission/PermissionManagerImpl.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2020 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package azkaban.imagemgmt.permission;
+
+import azkaban.imagemgmt.daos.ImageTypeDao;
+import azkaban.imagemgmt.exeception.ImageMgmtException;
+import azkaban.imagemgmt.models.ImageOwnership;
+import azkaban.imagemgmt.models.ImageOwnership.Role;
+import azkaban.imagemgmt.models.ImageType;
+import azkaban.user.Permission;
+import azkaban.user.Permission.Type;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.collections4.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class defines and manages the permission for accessing image management APIs. The
+ * permissions are defined on image type. The permissions are created using user role and access
+ * type. This class internally accesses image type cache to get image type metadata.
+ */
+@Singleton
+public class PermissionManagerImpl implements PermissionManager {
```

Comment:
Have you tried to reuse the existing GROUP-ROLE-PERMISSION mechanism that's implemented in our azkaban-users.xml file to add authorization to these new APIs? That way maybe we don't need to write new code and it's just a matter of defining a new "image-management" group/role in the xml file and then ask for membership to that group/role when creating a new image type or image version.

Reply:
The main issue with azkaban-user.xml we will be restricted to only headless account or service account. The idea of providing image management APIs is to provide access to individual LDAP user. Hence, LDAP user role and group assignment can't be done using azkaban-users.xml. Moreover, using API we can provision to add/deleve image owners. This is not possible if we maintain the permission part in the azkaban-users.xml and also it not feasible to modify the file dynamically. Hence, separate user mangement and permission for make more sense while managing image type owners. Moreover, we can add service account and image owner as well using API. Howerver, add and delete image owner separate API will be provided in the future. Currently while creating/registering the image type only we can specify owners.

Reply:
Actually the `azkaban-user.xml` file is used for more than just service accounts. For example, we can define:
```
<group name="image-management-users" roles="image-management" /> 
<role name="image-management" permissions=CRUD_IMAGE_MANAGEMENT_APIS" />
``` 
Then, on LDAP we create a group with same name and add to it all the users that can use the APIs.
When a new image management API request comes we check the user has appropriate permissions(CRUD_IMAGE_MANAGEMENT_APIS) which will only happen if they belong to the LDAP group.
In addition, for update and delete operations on existing image types, we would also need to check if the user is ADMIN or MEMBER of the specific image type.

I haven't tested this but it's a way to reuse existing GROUP-ROLE-PERMISSION mechanism without needing to create the PermissionManager and PermissionManagerImpl class hierarchy.